pallets/settlement/src/lib.rs

```diff
@@ -1118,22 +1266,103 @@ impl<T: Trait> Module<T> {
                 }
             }
 
+            enum FailureReason {
```

Comment:
We can declare this outside the function and create an `into` implementation for this to directly convert the reason into error.